File: python_code/Bootstrap_new/a4_bayesselect_variables.py
import numpy as np
import pandas as pd
import os


def a4_bayesselect_variables(GQ, X0, vars_list):
    base_dir = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
    save_dir = os.path.join(base_dir, 'saved_models')
    data = np.load(os.path.join(save_dir, 'RF_bayesian_predictor_result.npz'), allow_pickle=True)
    mytable = data['mytable']

    table1 = mytable[:20]
    table2 = mytable[20:40]
    table3 = mytable[40:60]
    table4 = mytable[60:80]
    table5 = mytable[80:100]
    table_list = [table1, table2, table3, table4, table5]

    ret_feature = []
    parameters = []

    for i in range(5):
        tab = sorted(table_list[i], key=lambda x: x['CVR2'], reverse=True)
        z = tab[0]
        ind = np.array([z[f'var{j+1}'] == 'true' for j in range(12)])
        para = {'MinLeafSize': z['MinLeafSize'], 'MaxNumSplits': z['MaxNumSplits'],
                'NumVariablesToSample': z['NumVariablesToSample'], 'NumLearningCycles': z['NumLearningCycles']}
        features = [vars_list[j] for j in range(12) if ind[j]]
        ret_feature.append(features)
        parameters.append(para)

    np.savez(os.path.join(save_dir, 'slected_features.npz'),
             ret_feature=np.array(ret_feature, dtype=object),
             parameters=np.array(parameters, dtype=object))
    print('')

File: Bootstrap_new/a4_bayesselect_variables.m
function a4_bayesselect_variables % Solving the system, check
close all
clearvars
clc
format short g, format compact
profile on
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
load b0_X_GQ.mat X0 GQ
X=X0{:,:};
load RF_bayesian_predictor_result.mat mytable

table1=mytable(1:20,:);table2=mytable(21:40 ,:); table3=mytable(41:60 ,:); table4=mytable(61:80 ,:); table5=mytable(81:100 ,:);
table={table1 table2 table3 table4 table5};

for i=1:5
tab=sortrows(table{i},"CVR2","descend");
z=tab(1,:);
ind=z{1,1:12}=='true';
para=z{1,17:18};
features=X0.Properties.VariableNames(:,ind);
ret_feature{i}= features;
parameters{i}=para;
end
save slected_features.mat ret_feature parameters
disp('')
